File: packages/expo-crypto/android/src/main/java/expo/modules/crypto/CryptoModule.kt
package expo.modules.crypto

import android.util.Base64
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.typedarray.TypedArray
import expo.modules.kotlin.typedarray.Uint8Array
import java.nio.charset.Charset
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.UUID
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

class CryptoModule : Module() {
  private val secureRandom by lazy { SecureRandom() }

  override fun definition() = ModuleDefinition {
    Name("ExpoCrypto")

    Class("AesGcmParams")

    Class(CryptoKey::class) {
      Constructor { algorithm:AesKeyGenParams, extractable: Boolean, keyUsages: List<KeyUsage> ->
        return@Constructor CryptoKey(algorithm, extractable, keyUsages)
      }
    }

    Function("digestString", this@CryptoModule::digestString)
    AsyncFunction("digestStringAsync", this@CryptoModule::digestString)
    Function("getRandomBase64String", this@CryptoModule::getRandomBase64String)
    AsyncFunction("getRandomBase64StringAsync", this@CryptoModule::getRandomBase64String)
    Function("getRandomValues", this@CryptoModule::getRandomValues)
    Function("digest", this@CryptoModule::digest)
    Function("randomUUID") {
      UUID.randomUUID().toString()
    }
    Function("encryptAesGcm", this@CryptoModule::encryptAesGcm)
    Function("decryptAesGcm", this@CryptoModule::decryptAesGcm)
    Function("exportKey", this@CryptoModule::exportKey)
    Function("importKey", this@CryptoModule::importKey)
  }

  private fun getRandomBase64String(randomByteCount: Int): String {
    val output = ByteArray(randomByteCount)
    secureRandom.nextBytes(output)
    return Base64.encodeToString(output, Base64.NO_WRAP)
  }

  private fun digestString(algorithm: DigestAlgorithm, data: String, options: DigestOptions): String {
    val messageDigest = MessageDigest.getInstance(algorithm.value).apply { update(data.toByteArray()) }

    val digest: ByteArray = messageDigest.digest()
    return when (options.encoding) {
      DigestOptions.Encoding.BASE64 -> {
        Base64.encodeToString(digest, Base64.NO_WRAP)
      }
      DigestOptions.Encoding.HEX -> {
        digest.joinToString(separator = "") { byte ->
          ((byte.toInt() and 0xff) + 0x100)
            .toString(radix = 16)
            .substring(startIndex = 1)
        }
      }
    }
  }

  private fun digest(algorithm: DigestAlgorithm, output: TypedArray, data: TypedArray) {
    val messageDigest = MessageDigest.getInstance(algorithm.value).apply { update(data.toDirectBuffer()) }

    val digest: ByteArray = messageDigest.digest()
    output.write(digest, output.byteOffset, output.byteLength)
  }

  private fun getRandomValues(typedArray: TypedArray) {
    val array = ByteArray(typedArray.byteLength)
    secureRandom.nextBytes(array)
    typedArray.write(array, typedArray.byteOffset, typedArray.byteLength)
  }

  private fun encryptAesGcm(key: CryptoKey, data: String, iv: Uint8Array): String {
    if(!key.keyUsages.contains(KeyUsage.ENCRYPT)) {
      throw WrongKeyUsageException();
    }

    val dataBytes = data.toByteArray();

    val ivBuffer = iv.toDirectBuffer();
    val ivBytes = ByteArray(iv.length).also (ivBuffer::get)

    val spec = GCMParameterSpec(128, ivBytes)

    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    val keySpec = SecretKeySpec(key.getKey().encoded,"AES")
    cipher.init(Cipher.ENCRYPT_MODE, keySpec, spec)

    val cipherText = cipher.doFinal(dataBytes)

    return Base64.encodeToString(cipherText, Base64.DEFAULT)
  }

  private fun decryptAesGcm(key: CryptoKey, data: String, iv: Uint8Array): String {
    if(!key.keyUsages.contains(KeyUsage.DECRYPT)) {
      throw WrongKeyUsageException();
    }

    val ivBuffer = iv.toDirectBuffer();
    val ivBytes = ByteArray(iv.length).also(ivBuffer::get)

    val spec = GCMParameterSpec(128, ivBytes)
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    val keySpec = SecretKeySpec(key.getKey().encoded,"AES")
    cipher.init(Cipher.DECRYPT_MODE, keySpec, spec)

    val decoded = Base64.decode(data, Base64.DEFAULT)

    val plaintext = cipher.doFinal(decoded)

    return String(plaintext, Charset.forName("UTF-8"))
  }

  private fun exportKey(format: KeyFormat, key: CryptoKey, dest: Uint8Array) {
    if(format != KeyFormat.RAW) {
      throw WrongKeyUsageException();
    }

    //  TODO check for size
    val keyBytes = key.getKey().encoded
    dest.write(keyBytes, position = 0, size = keyBytes.size)
  }

  private fun importKey(format: KeyFormat, key: Uint8Array, cryptoKey: CryptoKey) {
    if(format != KeyFormat.RAW) {
      throw WrongKeyUsageException();
    }

    //  TODO check for size
    val secKey = key.toSecretKey()
    cryptoKey.setKey(secKey)
  }
}

fun ByteArray.toSecretKey(algorithm: String = "AES") =
  SecretKeySpec(this, 0, this.size, algorithm)

fun Uint8Array.toSecretKey(): SecretKey {
  return ByteArray(32)
    .also { bytes -> this.read(bytes, 0, 32) }
    .toSecretKey()
}

class WrongKeyUsageException : CodedException(
  message = "Wrong key usage"
)

class InvalidKeyAlgorithm : CodedException(
  message = "Invalid key Algorithm"
)

File: packages/expo-crypto/ios/CryptoModule.swift
// Copyright 2022-present 650 Industries. All rights reserved.

import CommonCrypto
import ExpoModulesCore
import CryptoKit

public class CryptoModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoCrypto")

    Class(CryptoKey.self) {
        Constructor { (algorithm: AesKeyGenParams, extractable: Bool, keyUsages: [KeyUsage]) in
            return CryptoKey(algorithm: algorithm, extractable: extractable, keyUsages: keyUsages)
        }
    }

    AsyncFunction("digestStringAsync", digestString)

    Function("digestString", digestString)

    AsyncFunction("getRandomBase64StringAsync", getRandomBase64String)

    Function("getRandomBase64String", getRandomBase64String)

    Function("getRandomValues", getRandomValues)

    Function("digest", digest)

    Function("randomUUID") {
      UUID().uuidString.lowercased()
    }

    Function("encryptAesGcm", encryptAesGcm)

    Function("decryptAesGcm", decryptAesGcm)

    Function("exportKey", exportKey)

   Function("importKey", importKey)

  }
}

internal enum KeyFormat: String, EnumArgument {
  case raw = "raw"
  case pkcs8 = "pkcs8"
  case spki = "spki"
  case jwk = "jwk"
}

internal enum KeyUsage: String, EnumArgument {
  case encrypt = "encrypt"
  case decrypt = "decrypt"
  case sign = "sign"
  case verify = "verify"
  case deriveKey = "deriveKey"
  case deriveBits = "deriveBits"
  case wrapKey = "wrapKey"
  case unwrapKey = "unwrapKey"
}

struct AesKeyGenParams: Record {
  @Field
  var name: AlgorithmName = .gcm

  @Field
  var length: Int = 32

  internal enum AlgorithmName: String, EnumArgument {
    case cbc = "AES-CBC"
    case gcm = "AES-GCM"
  }
}

struct HmacKeyGenParams: Record {
  @Field
  var name: AlgorithmName = .hmac

  internal enum AlgorithmName: String, EnumArgument {
    case hmac = "HMAC"
  }
}

class CryptoKey: SharedObject {
  private var _algorithm: AesKeyGenParams
  private var _extractable: Bool
  private let _keyUsages: [KeyUsage]

  private var algorithm: AesKeyGenParams {
    _algorithm
  }
  private var extractable: Bool {
    _extractable
  }
  var keyUsages: [KeyUsage] {
    _keyUsages
  }

  var secretKey = SymmetricKey(size: .bits256)

  init(algorithm: AesKeyGenParams, extractable: Bool, keyUsages: [KeyUsage]) {
    self._algorithm = algorithm
    self._extractable = extractable
    self._keyUsages = keyUsages
    if algorithm.name == .gcm {
      secretKey = SymmetricKey(size: .bits256)
    }
  }

  func getKey() -> SymmetricKey {
    return secretKey
  }

  func setKey(newKey: SymmetricKey) {
    self.secretKey = newKey
  }

}


private func getRandomBase64String(length: Int) throws -> String {
  var bytes = [UInt8](repeating: 0, count: length)
  let status = SecRandomCopyBytes(kSecRandomDefault, length, &bytes)

  guard status == errSecSuccess else {
    throw FailedGeneratingRandomBytesException(status)
  }
  return Data(bytes).base64EncodedString()
}

private func digestString(algorithm: DigestAlgorithm, str: String, options: DigestOptions) throws -> String {
  guard let data = str.data(using: .utf8) else {
    throw LossyConversionException()
  }

  let length = Int(algorithm.digestLength)
  var digest = [UInt8](repeating: 0, count: length)

  data.withUnsafeBytes { bytes in
    let _ = algorithm.digest(bytes.baseAddress, UInt32(data.count), &digest)
  }

  switch options.encoding {
  case .hex:
    return digest.reduce("") { $0 + String(format: "%02x", $1) }
  case .base64:
    return Data(digest).base64EncodedString()
  }
}

private func getRandomValues(array: TypedArray) throws -> TypedArray {
  let status = SecRandomCopyBytes(
    kSecRandomDefault,
    array.byteLength,
    array.rawPointer
  )

  guard status == errSecSuccess else {
    throw FailedGeneratingRandomBytesException(status)
  }
  return array
}

private func digest(algorithm: DigestAlgorithm, output: TypedArray, data: TypedArray) {
  let length = Int(algorithm.digestLength)
  let outputPtr = output.rawPointer.assumingMemoryBound(to: UInt8.self)
  algorithm.digest(data.rawPointer, UInt32(data.byteLength), outputPtr)
}

private func encryptAesGcm(key: CryptoKey, data: String, iv: Uint8Array) throws -> String {
  guard key.keyUsages.contains(.encrypt) else {
    throw WrongKeyUsageException()
  }

  //TODO handle invalid length
  let nonceData = iv.data()
  let nonce = try AES.GCM.Nonce(data: nonceData)

  guard let plaintextData = data.data(using: .utf8) else {
    throw WrongKeyUsageException()  //TOOO improve error handling
  }

  let encryptionResult = try AES.GCM.seal(
    plaintextData,
    using: key.getKey(),
    nonce: nonce)

  //TODO this should use  ciphertext but we need to return also TAG
  let cipherText = encryptionResult.combined

  guard let cipherText = encryptionResult.combined else {
    throw EncryptionFailedException("Incorrect AES configuration")
  }

  let base64cipherText = cipherText.base64EncodedString()
  return base64cipherText
}

private func decryptAesGcm(key: CryptoKey, data: String, iv: Uint8Array) throws -> String {
  guard key.keyUsages.contains(.encrypt) else {
    throw WrongKeyUsageException()
  }

  //TODO handle invalid length
  let nonceData = iv.data()
  let nonce = try AES.GCM.Nonce(data: nonceData)  // not needed but we need to migrate to using TAG

  guard let cipherData = Data(base64Encoded: data) else {
    throw WrongKeyUsageException()  // TODO: Improve error handling
  }

  let sealedBox = try AES.GCM.SealedBox(combined: cipherData)
  let plaintextData = try AES.GCM.open(sealedBox, using: key.getKey())

  guard let plaintext = String(data: plaintextData, encoding: .utf8) else {
    throw WrongKeyUsageException()  // TODO: Improve error handling
  }

  return plaintext
}

private func exportKey(format: KeyFormat, key: CryptoKey, dest: Uint8Array) throws {
  guard format == .raw else {
    throw WrongKeyUsageException()
  }

  //  TODO check for size
  key.getKey().withUnsafeBytes { bytes in
    let _ = bytes.copyBytes(to: dest.rawBufferPtr())
  }
}

private func importKey(format: KeyFormat, key: Uint8Array) throws -> CryptoKey {
  guard format == .raw else {
    throw WrongKeyUsageException()
  }

  let key = SymmetricKey(data: key.data())
  let cryptoKey = CryptoKey(
    algorithm: AesKeyGenParams(name: .gcm, length: 32), extractable: true,
    keyUsages: [.encrypt, .decrypt])
  cryptoKey.setKey(newKey: key)
  return cryptoKey
}



extension TypedArray {
  func data() -> Data {
    Data(bytes: self.rawPointer, count: self.byteLength)
  }

  func rawBufferPtr() -> UnsafeMutableRawBufferPointer {
    UnsafeMutableRawBufferPointer(
      start: self.rawPointer,
      count: self.byteLength)
  }
}

private class InvalidNonceException: Exception {
  override var reason: String {
    "Invalid nonce"
  }
}

private class WrongKeyUsageException: Exception {
  override var reason: String {
    "Wrong key usage"
  }
}

private class EncryptionFailedException: GenericException<String> {
  override var reason: String {
    "Failed to encrypt data: \(param)"
  }
}
private class LossyConversionException: Exception {
  override var reason: String {
    "Unable to convert given string without losing some information"
  }
}

private class FailedGeneratingRandomBytesException: GenericException<OSStatus> {
  override var reason: String {
    "Generating random bytes has failed with OSStatus code: \(param)"
  }
}
